File: python/google-adk/sample-agent/mcp_tool_registration_service.py
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.

from typing import Optional
import logging

from google.adk.agents import Agent
from google.adk.tools.mcp_tool.mcp_toolset import McpToolset, StreamableHTTPConnectionParams

from microsoft_agents.hosting.core import Authorization, TurnContext

from microsoft_agents_a365.runtime.utility import Utility
from microsoft_agents_a365.tooling.models import ToolOptions
from microsoft_agents_a365.tooling.services.mcp_tool_server_configuration_service import (
    McpToolServerConfigurationService,
)
from microsoft_agents_a365.tooling.utils.constants import Constants
from microsoft_agents_a365.tooling.utils.utility import (
    get_mcp_platform_authentication_scope,
)


class McpToolRegistrationService:
    """Service for managing MCP tools and servers for an agent"""

    _orchestrator_name: str = "GoogleADK"

    def __init__(self, logger: Optional[logging.Logger] = None):
        """
        Initialize the MCP Tool Registration Service for Google ADK.

        Args:
            logger: Logger instance for logging operations.
        """
        self._logger = logger or logging.getLogger(self.__class__.__name__)
        self.config_service = McpToolServerConfigurationService(logger=self._logger)

    async def add_tool_servers_to_agent(
        self,
        agent: Agent,
        agentic_app_id: str,
        auth: Authorization,
        auth_handler_name: str,
        context: TurnContext,
        auth_token: Optional[str] = None,
    ):
        """
        Add new MCP servers to the agent by creating a new Agent instance.

        Note: This method creates a new Agent instance with MCP servers configured.

        Args:
            agent: The existing agent to add servers to.
            agentic_app_id: Agentic App ID for the agent.
            auth: Authorization object used to exchange tokens for MCP server access.
            auth_handler_name: Name of the authorization handler.
            context: TurnContext object representing the current turn/session context.
            auth_token: Authentication token to access the MCP servers. If not provided,
                        will be obtained using `auth` and `context`.

        Returns:
            New Agent instance with all MCP servers
        """
        # Acquire auth token if not provided
        if not auth_token:
            scopes = get_mcp_platform_authentication_scope()
            auth_token_obj = await auth.exchange_token(context, scopes, auth_handler_name)
            auth_token = auth_token_obj.token

        self._logger.info(f"Listing MCP tool servers for agent {agentic_app_id}")

        options = ToolOptions(orchestrator_name=self._orchestrator_name)

        # Pass auth context for V2 per-audience token acquisition (production path).
        # In dev/Playground mode (empty auth_handler_name), the SDK reads per-server
        # tokens from BEARER_TOKEN_MCP_<SERVER> / BEARER_TOKEN env vars automatically
        # via its internal _attach_dev_tokens method.
        list_kwargs = {}
        if auth_handler_name:
            list_kwargs = {
                "authorization": auth,
                "auth_handler_name": auth_handler_name,
                "turn_context": context,
            }

        mcp_server_configs = await self.config_service.list_tool_servers(
            agentic_app_id=agentic_app_id,
            auth_token=auth_token,
            options=options,
            **list_kwargs,
        )

        self._logger.info(f"Loaded {len(mcp_server_configs)} MCP server configurations")

        # Convert MCP server configs to McpToolset objects
        mcp_servers_info = []

        for server_config in mcp_server_configs:
            if not server_config.url:
                self._logger.warning(
                    "Skipping MCP server '%s' — no URL configured (dev mode or manifest-only config).",
                    server_config.mcp_server_unique_name,
                )
                continue

            # server_config.headers already contains the per-audience Authorization token:
            # - Dev mode:  set by SDK's _create_dev_token_acquirer (reads BEARER_TOKEN_MCP_* / BEARER_TOKEN)
            # - Prod mode: set by SDK's _create_obo_token_acquirer (per-audience OBO exchange)
            base_headers = {
                Constants.Headers.USER_AGENT: Utility.get_user_agent_header(
                    self._orchestrator_name
                )
            }
            server_level_headers = dict(server_config.headers) if server_config.headers else {}
            mcp_server_headers = {**base_headers, **server_level_headers}

            has_auth = Constants.Headers.AUTHORIZATION in mcp_server_headers
            self._logger.info(
                "Configuring MCP server '%s' → %s (auth_header=%s)",
                server_config.mcp_server_name,
                server_config.url,
                "present" if has_auth else "MISSING — MCP calls will fail without a valid token",
            )

            server_info = McpToolset(
                connection_params=StreamableHTTPConnectionParams(
                    url=server_config.url,
                    headers=mcp_server_headers,
                    timeout=30.0,
                )
            )

            mcp_servers_info.append(server_info)

        all_tools = agent.tools + mcp_servers_info

        return Agent(
            name=agent.name,
            model=agent.model,
            description=agent.description,
            instruction=agent.instruction,
            tools=all_tools,
        )

File: nodejs/google-adk/sample-agent/mcpToolRegistrationService.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import { Agent, MCPToolset } from '@google/adk';

import {
  McpToolServerConfigurationService,
  ToolingConfiguration,
  resolveTokenScopeForServer,
} from '@microsoft/agents-a365-tooling';

import type { TurnContext } from '@microsoft/agents-hosting';

// Use axios directly to call the gateway (same as the SDK uses internally)
import axios from 'axios';

const logger = {
  info: (...args: unknown[]) =>
    console.log(new Date().toISOString(), 'INFO', 'McpToolRegistrationService:', ...args),
  warn: (...args: unknown[]) =>
    console.warn(new Date().toISOString(), 'WARN', 'McpToolRegistrationService:', ...args),
  error: (...args: unknown[]) =>
    console.error(new Date().toISOString(), 'ERROR', 'McpToolRegistrationService:', ...args),
};

export interface AddToolServersOptions {
  agent: Agent;
  agenticAppId: string;
  auth: unknown;
  authHandlerName: string | null;
  context: TurnContext;
  authToken?: string;
}

export class McpToolRegistrationService {
  private configService: McpToolServerConfigurationService;

  constructor() {
    this.configService = new McpToolServerConfigurationService();
  }

  /**
   * Add new MCP servers to the agent by creating a new Agent instance.
   */
  async addToolServersToAgent(options: AddToolServersOptions): Promise<Agent> {
    const { agent, agenticAppId, auth, authHandlerName, context, authToken: providedToken } = options;

    let authToken = providedToken;

    if (!authToken && auth && authHandlerName) {
      // Exchange token using the authorization object with the MCP platform scope.
      // The scope comes from ToolingConfiguration (ea9ffc3e-.../.default),
      // NOT https://api.powerplatform.com/.default.
      try {
        const authObj = auth as any;
        if (typeof authObj.exchangeToken === 'function') {
          const mcpScope = new ToolingConfiguration().mcpPlatformAuthenticationScope;
          logger.info(`Exchanging token via auth handler '${authHandlerName}' for MCP scope: ${mcpScope}`);
          const authTokenObj = await authObj.exchangeToken(context, authHandlerName, {
            scopes: [mcpScope],
          });
          authToken = authTokenObj?.token;
          logger.info(`Token exchange result: ${authToken ? `success (length: ${authToken.length})` : 'null/empty'}`);
        } else {
          logger.warn('auth object does not have exchangeToken method');
        }
      } catch (err) {
        logger.error('Token exchange failed:', err);
        return agent;
      }
    }

    if (!authToken) {
      logger.warn('No auth token available for MCP tool servers');
      return agent;
    }

    logger.info(`Listing MCP tool servers for agent: '${agenticAppId}'`);

    let mcpServerConfigs: any[];
    try {
      // Call the A365 tooling gateway directly to handle response shape variations.
      // The SDK's listToolServers expects response.data to be an array, but the
      // gateway may return { mcpServers: [...] } (object with array property).
      const toolingConfig = new ToolingConfiguration();
      const endpoint = `${toolingConfig.mcpPlatformEndpoint}/agents/v2/${agenticAppId}/mcpServers`;
      logger.info(`Gateway URL: ${endpoint}`);

      const response = await axios.get(endpoint, {
        headers: { Authorization: `Bearer ${authToken}` },
        timeout: 10000,
      });

      logger.info(`Gateway response status: ${response.status}`);
      logger.info(`Gateway response type: ${typeof response.data}, isArray: ${Array.isArray(response.data)}`);

      // Handle both shapes: raw array OR { mcpServers: [...] }
      const rawServers = Array.isArray(response.data)
        ? response.data
        : Array.isArray(response.data?.mcpServers)
          ? response.data.mcpServers
          : [];

      mcpServerConfigs = rawServers.map((s: any) => ({
        mcpServerName: s.mcpServerName,
        mcpServerUniqueName: s.mcpServerUniqueName ?? s.mcpServerName,
        url: s.url,
        headers: s.headers,
        audience: s.audience,
        scope: s.scope,
        publisher: s.publisher,
      }));

      logger.info(`Loaded ${mcpServerConfigs.length} MCP server configurations`);
      for (const cfg of mcpServerConfigs) {
        logger.info(`  Server: ${cfg.mcpServerUniqueName ?? '(unknown)'}, URL: ${cfg.url ?? '(none)'}`);
      }
    } catch (err: any) {
      logger.error(`Failed to list MCP tool servers:`);
      logger.error(`  agenticAppId: '${agenticAppId}'`);
      logger.error(`  Error: ${err.message}`);
      if (err.response) {
        logger.error(`  HTTP Status: ${err.response.status}`);
        logger.error(`  Response data: ${JSON.stringify(err.response.data)}`);
      }
      throw err;
    }

    // Convert MCP server configs to MCPToolset objects
    const mcpServersInfo: MCPToolset[] = [];

    for (const serverConfig of mcpServerConfigs) {
      if (!serverConfig.url) {
        logger.warn(
          `Skipping MCP server '${serverConfig.mcpServerUniqueName}' — no URL configured.`
        );
        continue;
      }

      // MCPToolset requires connectionParams with:
      //   type: "StreamableHTTPConnectionParams" (discriminant for the switch)
      //   url: the server endpoint
      //   header: auth headers (note: singular "header", not "headers")
      const serverInfo = new MCPToolset({
        type: 'StreamableHTTPConnectionParams',
        url: serverConfig.url,
        header: { Authorization: `Bearer ${authToken}` },
      } as any);

      logger.info(`Created MCPToolset for '${serverConfig.mcpServerUniqueName}' at ${serverConfig.url}`);
      mcpServersInfo.push(serverInfo);
    }

    const existingTools = agent.tools ?? [];
    const allTools = [...existingTools, ...mcpServersInfo];

    return new Agent({
      name: agent.name,
      model: agent.model as string,
      description: agent.description,
      instruction: agent.instruction as string,
      tools: allTools,
    });
  }
}
